### ISSUE TITLE
Static executables on other platforms: Cross compilation via Nix

### ISSUE THREAD

**Issue by wolfgangwalther** [PostgREST/postgrest#3281]
We have been building a static executable for x64-linux for quite a while now. From this, the minimal docker image is created.

It would be really cool, if we could build the static executable on other platforms, too - and then create a multi-layer docker image, which provides minimal image size for different platforms.

One way to do this would be to run a nix environment on linux x64 and build the static executable for other platforms via **cross-compiling**.

This issue collects the current state of affairs.

---

I have been building GHC 9.4.8 as a cross-compiler for aarch64 successfully. However, when compiling PostgREST and its dependencies a major challenge comes up quickly: Template Haskell.

Template Haskell is a challenge, because it needs to be **run during compilation**. Because the cross-compiler compiles for the target platform, the compiled TH code can't be executed on the build platform. GHC will then fail like this:

```
Couldn't find a target code interpreter. Try with -fexternal-interpreter
```

There are a few different possible approaches how to deal with that:
1. Remove all Template Haskell from PostgREST and its dependencies. This is not easy, because a few transitive dependencies use TH and for example the whole hasql ecosystem depends on them. We also use it ourselves for the `Paths_` module, although that could probably be worked around.
2. Compile and execute TH natively and then pass the resulting splices to the cross-compiler. The idea is, that most Template Haskell applications are merely "code generation", e.g. helpers to create a lot of similar shaped haskell code, which is then compiled to the actual program. Many of those uses are actually independent of the target system - so it could be possible to compile and execute those splices on the build system and then pass the generated code to the cross-compiler. I experimented with that and I was able to cross-compile almost all our dependencies that way. I was doing it manually and this quickly got very annoying, that's why I didn't finish this, yet. This could possibly be automated, but it would never be the "clean" solution. Once any of our dependencies would use "unsafe" Template Haskell, this might result in hard to debug errors.
3. GHC comes with a `ghc-iserv` executable. This is basically a wrapper around the `GHCi` library - so it allows to execute compiled code. For a GHC cross-compiler, this executable is built for the target platform. This can therefore be run with `qemu` and passed to `-fexternal-interpreter`. This should make Template Haskell just work. I tried this with GHC 9.4.8 and was able to make it work - but at some point hit a compiler bug in GHC 9.4.8 when cross-compiling to aarch64. My understanding is, that this bug was fixed in GHC 9.6 and 9.8 - but not in the 9.4 series. Therefore, I tried upgrading to GHC 9.6 / 9.8, but couldn't make that work, yet. GHC 9.6 doesn't build for `pkgsStatic` in nixpkgs, yet - and GHC 9.8 does build for `pkgsStatic`, but fails with the same TH-related error message as mentioned above, but already for `pkgsStatic` (no cross-platform-compiling involved, yet). This is tracked in https://github.com/NixOS/nixpkgs/issues/275304.

The last approach is the most promising in principle and is also tracked in https://github.com/NixOS/nixpkgs/issues/248979 upstream.

**Comment by alexfmpe** [comment]
@wolfgangwalther I think this is now supported in nixpkgs right?

**Comment by wolfgangwalther** [comment]
Yes, didn't check but this should have reached master by now. Will have to update our nixpkgs pin - it's on my (long) list of things to do.

**Comment by wolfgangwalther** [comment]
Finally updated the Nixpkgs pin. Since we're building aarch64-linux natively now, we don't need cross for it. We can't cross-compile to FreeBSD with iserv-proxy, because AFAIK there is no qemu / emulator for it. The iserv-proxy changes in Nixpkgs will most likely be useful to do a Windows build in the future, but we will need to solve a few other things on the Nixpkgs-side for that first - notably cross-compiling `libpq`.

It would still be best to remove our dependency on TemplateHaskell, because that would allow us to cross-compile to FreeBSD once this builds in Nixpkgs.

**Comment by steve-chavez** [comment]
@wolfgangwalther I think we should put the new static "distroless" aarch64 release ([ref](https://github.com/PostgREST/postgrest/commit/fee820bac28dc00380007c17a6c09f9035252db7)) somewhere in the changelog, it feels like a feature. It could also be a breaking change, now that I think about it.
